dhall-docs/src/Dhall/Docs/Comment.hs

```diff
@@ -0,0 +1,305 @@
+            ||  c == '\n'
+            ||  c == '\t'
+
+    character = (Combinators.satisfy predicate)
+      where
+        predicate c = '\x20' <= c && c <= '\x10FFFF' || c == '\n' || c == '\t'
+
+    endOfLine = (Token.text "\r\n" <?> "newline")
+
+blockCommentContinue :: Parser Text
+blockCommentContinue = endOfComment <|> continue
+  where
+    endOfComment = void (Token.text "-}") *> pure ""
+
+    continue = do
+        c <- blockCommentChunk
+        c' <- blockCommentContinue
+        pure (c <> c')
+
+-- | Consume whitespace lines or lines that only have whitespaces *before* a comment
+whitespace :: Parser ()
+whitespace = Text.Megaparsec.skipMany (Text.Megaparsec.choice
+    [ void (Combinators.takeWhile1 predicate)
+    , void (Token.text "\r\n")
+    ] <?> "whitespace")
+  where
+    predicate c = c == ' ' || c == '\t' || c == '\n'
+
+blockCommentParser :: Parser (DhallComment RawComment)
+blockCommentParser = do
+    c <- blockCommentParser'
+    whitespace
+    pure $ BlockComment ("{-" <> c <> "-}")
+
+-- | Parse all comments in a text fragment
+parseComments :: String -> Text -> [DhallComment RawComment]
+parseComments delta text = case result of
+    Left err -> error ("An error has occurred while parsing comments:\n "
+      <> Text.Megaparsec.errorBundlePretty err)
+    Right comments -> comments
+  where
+    parser = do
+        comments <- many $ do
+            whitespace
+            lineCommentParser <|> ((:| []) <$> blockCommentParser)
+        Text.Megaparsec.eof
+        pure $ concatMap NonEmpty.toList comments
+
+    result = Text.Megaparsec.parse (unParser parser) delta text
+
+data CommentParseError
+    = MissingNewlineOnBlockComment
+    | BadSingleLineCommentsAlignment
+    | BadPrefixesOnSingleLineComments
+    | SeveralSubseqDhallDocsComments
+    deriving Show
+
+-- | Checks if a 'RawComment' has the @dhall-docs@ marker
+parseMarkedComment :: DhallComment RawComment -> Maybe (DhallComment MarkedComment)
+parseMarkedComment (BlockComment comment)
+    | "{-|" `Data.Text.isPrefixOf` comment = Just $ BlockComment comment
+    | otherwise = Nothing
+
+parseMarkedComment (SingleLineComments ls)
+    | any (("--|" `Data.Text.isPrefixOf`) . snd) ls = Just (SingleLineComments ls)
+    | otherwise = Nothing
+
+-- | Knowning that there is a @dhall-docs@ marker inside the comment, this
+--   checks if a 'MarkedComment' is a 'DhallDocsComment'. For 'SingleLineComments'
+--   this also removes the prefix lines before the first marked comment
+parseDhallDocsComment :: DhallComment MarkedComment -> Either CommentParseError (DhallComment DhallDocsComment)
+parseDhallDocsComment (BlockComment comment) =
+    BlockComment <$> checkNewlineAfterMarker comment
+  where
+    checkNewlineAfterMarker :: Text -> Either CommentParseError Text
+    checkNewlineAfterMarker t =
+        if any (`Data.Text.isPrefixOf` t) ["{-|\n", "{-|\r\n"] then pure t
+        else Left MissingNewlineOnBlockComment
```

Comment:
This can be simplified a little bit:

```suggestion
    if any (`Data.Text.isPrefixOf` t) ["{-|\n", "{-|\r\n"]
    then BlockComment t
    else Left MissingNewlineOnBlockComment
```